src/animated/Controller.ts

```diff
@@ -4,61 +4,84 @@ import {
   is,
   toArray,
   withDefault,
+  hasKeys,
 } from '../shared/helpers'
 import AnimatedValue from './AnimatedValue'
 import AnimatedValueArray from './AnimatedValueArray'
 import { start, stop } from './FrameLoop'
 import { colorNames, interpolation as interp, now } from './Globals'
+import { SpringProps } from '../../types/renderprops'
 
-type FinishedCallback = (finished?: boolean) => void
+type Animation = any
+
+type ControllerProps<DS extends object> = DS &
+  SpringProps<DS> & {
+    attach?: (ctrl: Controller) => Animation
+  }
 
 type AnimationsFor<P> = { [Key in keyof P]: any }
 
 type ValuesFor<P> = { [Key in keyof P]: any }
 
-type InterpolationsFor<P> = {
-  [Key in keyof P]: P[Key] extends ArrayLike<any>
+type InterpolationsFor<DS> = {
+  [K in keyof DS]: DS[K] extends ArrayLike<any>
     ? AnimatedValueArray
     : AnimatedValue
 }
 
-let G = 0
-class Controller<P extends any = {}> {
-  id: number
+type EndCallback = (finished?: boolean) => void
 
+let nextId = 1
+class Controller<DS extends object = any> {
+  id = nextId++
   idle = true
-  hasChanged = false
   guid = 0
-  local = 0
-  props: P = {} as P
+  props: ControllerProps<DS> = {} as any
   merged: any = {}
-  animations = {} as AnimationsFor<P>
-  interpolations = {} as InterpolationsFor<P>
-  values = {} as ValuesFor<P>
+  animations = {} as AnimationsFor<DS>
+  interpolations = {} as InterpolationsFor<DS>
+  values = {} as ValuesFor<DS>
   configs: any = []
-  listeners: FinishedCallback[] = []
   queue: any[] = []
-  localQueue?: any[]
+  prevQueue: any[] = []
+  onEndQueue: EndCallback[] = []
+  pendingCount = 0
+
+  // Add this controller to the frameloop
+  private _start(onEnd?: EndCallback) {
+    if (onEnd) this.onEndQueue.push(onEnd)
+    if (this.idle) {
+      this.idle = false
+      start(this)
+    }
+  }
 
-  constructor() {
-    this.id = G++
+  private _stop(finished?: boolean) {
+    this.idle = true
+    stop(this)
+    if (finished && this.props.onRest) {
+      this.props.onRest(this.merged)
+    }
+    if (this.onEndQueue.length) {
+      this.onEndQueue.forEach(onEnd => onEnd(finished))
+      this.onEndQueue = []
+    }
   }
 
   /** update(props)
    *  This function filters input props and creates an array of tasks which are executed in .start()
    *  Each task is allowed to carry a delay, which means it can execute asnychroneously */
-  update(args?: P) {
-    //this._id = n + this.id
-
-    if (!args) return this
```

Comment:
I didn't see anywhere in the codebase that called `update` with an argument that's potentially falsy, so I removed this line. LMK if I should revert.

Reply:
that was just for debug purposes